components/automate-deployment/pkg/converge/converger.go

```diff
@@ -127,63 +136,460 @@ func StartConverger(opts ...ConvergerOpt) Converger {
+	topology := make(Topology)
+	topology[tgt] = []Service{
+		{
+			Name:          deploymentServiceName,
+			ConvergeState: Skip(),
+		},
 	}
+	desiredState := NewDesiredState(topology,
+		NewSkipSupervisorState(),
+		[]habpkg.HabPkg{}, // ignore-list is empty because cleanup mode is disabled
+		depot.DisabledGC)
+
+	return converger.sendMessage(t, &stopRequest{
+		eventSink:    eventSink,
+		desiredState: desiredState, //TODO(jaym) make a copy
+		doneChan:     t.C,
+	})
 }
 
 // Stop stops the converger
 func (converger *converger) Stop() {
 	converger.stop <- struct{}{}
 }
 
-func (converger *converger) run() {
-	for {
-		select {
-		case <-converger.stop:
-			//TODO(jaym) drain inbox and fail pending requests
-			close(converger.inbox)
-			close(converger.stop)
-			return
-		case req := <-converger.inbox:
-			if req.version < converger.lastVersion {
-				//TODO(jaym) send fail
-				logrus.WithFields(logrus.Fields{
-					"last_version":    converger.lastVersion,
-					"request_version": req.version,
-				}).Info("Skipping converge")
-			} else {
-				convergePlan, err := converger.compiler.Compile(req.desiredState)
-				if err != nil {
-					logrus.WithError(err).Error("Failed to compile")
-				} else {
-					err = convergePlan.Execute(req.eventSink)
-					if err != nil {
-						if err != api.ErrSelfUpgradePending {
-							logrus.WithError(err).Error("Converge failed")
-						}
-					}
-					converger.lastVersion = req.version
-				}
-				req.doneChan <- err
-			}
-			close(req.doneChan)
+func (converger *converger) sendMessage(t *Task, msg message) error {
+	select {
+	case converger.inbox <- msg:
+		logrus.WithField("converger_inbox_len", len(converger.inbox)).Debugf("Sent %s request to converger", msg)
+		return nil
+	default:
+		if t != nil {
+			close(t.C)
 		}
+		return errors.New("Inbox full")
+	}
+}
+
+// Debug
+//
+// Our converger has a debug channel where we can send debug requests
+// to query the current state of the converger. This is used in the
+// test. An alternative approach here would be to add a mutex.
+//
+
+type debugReq interface {
+	Do(*converger)
+}
+
+type debugResp struct {
+	state    state
+	inboxLen int
+}
+
+type debugGetReq struct {
+	responseChan chan debugResp
+}
+
+func (d debugGetReq) Do(c *converger) {
+	d.responseChan <- debugResp{
+		state:    copyState(c.currentState),
+		inboxLen: len(c.inbox),
+	}
+}
+
+type debugSetReq struct {
+	newState state
+	doneChan chan struct{}
+}
+
+func (d debugSetReq) Do(c *converger) {
+	logrus.WithFields(logrus.Fields{
+		"new_state": d.newState,
+	}).Debug("setting state via converger debug message")
+	c.setState(copyState(d.newState), false)
+	close(d.doneChan)
+}
+
+func (converger *converger) debugGetState() state {
+	respChan := make(chan debugResp, 1)
+	converger.debug <- debugGetReq{
+		responseChan: respChan,
+	}
+	resp := <-respChan
+	return resp.state
+}
+
+func (converger *converger) debugSetState(newState state) {
+	done := make(chan struct{})
```

Comment:
I'm going to call this a nit because it's test only code atm, but we should probably give up if the channel is full. Or use 1 channel

Reply:
By "use 1 channel" do you mean that these debug messages should go to `converger.inbox`? If so, then I definitely agree as that has a number of benefits.  I didn't do that here just because this was a late addition and I didn't want to clutter up the production code path.

I'm not sure I follow why we would want to give up on the debug message though?

Reply:
i think it's fine. I was just thinking of the case where we have multiple debug messages waiting just to get serviced and we're in a long converge loop, but i think it's fine since we service them first